compiler/rustc_parse/src/parser/nonterminal.rs

```diff
@@ -42,6 +42,8 @@ impl<'a> Parser<'a> {
                 && !token.is_keyword(kw::Let)
                 // This exception is here for backwards compatibility.
                 && !token.is_keyword(kw::Const)
+                // FIXME(vincenzopalazzo): query less time the edition
+                && (token.span.edition().at_least_rust_2024() && !token.is_keyword(kw::Underscore) )
```

Comment:
This doesn't seem correct -- `_` should not parse as an expression with `expr_2021` even if the caller edition is 2024.

Reply:
Otherwise this code will regress when upgraded from edition 2021 to 2024:

```
macro_rules! test {
    ($e:expr_2021) => { compile_error!("shouldn't be an expression") };
    (_) => {};
}

test!(_);
```

Reply:
I believe the the correct approach here is to default the `expr` macro fragment matcher to `expr_2021` in edition 2021 and the "real" `expr` matcher in 2024, similar to how `pat` defaults to `PatParam` on edition 2018 and `PatWithOr` in edition 2024.

See how `PatWithOr` is handled in `NonterminalKind::from_symbol`.

@eholk: This seems to be pre-existing since this PR was modeled after #123865, though. Why did you implement it this way rather than making it work like `PatWithOr`?

Reply:
Anyways, this kinda doesn't matter here -- I opened up a proposed fix at #126700 which should also simplify this PR as well.

Reply:
Thanks for the review, idk I wrote it a lot time ago, it is not ready for review, reason for be a draft.

But thanks to look into it, this will simplify my work in this week